tests/test_nvfuser.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+import logging
+
+import pytest
+import torch
+import torch.nn as nn
+from torch.cuda.amp.autocast_mode import autocast
+
+import xformers
+from xformers.components import Activation, LayerNormStyle
+from xformers.components.feedforward import build_feedforward
+
+_gpu_available = torch.cuda.is_available()
+
+xformers._is_functorch_available = True
+
+try:
+    from xformers.components.nvfuser import (
+        NVFusedBiasActivationDropout,
+        NVFusedBiasDropoutRes,
+        NVFusedBiasDropoutResLayerNorm,
+    )
+    from xformers.components.nvfuser.utils import build_nvfused
+except ImportError as e:
+    logging.warning(f"Functorch is not available in test_nvfuser.py. \nError {e}")
+
+FUSED_PATTERNS = (
+    [
+        NVFusedBiasActivationDropout,
+        NVFusedBiasDropoutRes,
+        NVFusedBiasDropoutResLayerNorm,
+    ]
+    if xformers._is_functorch_available
+    else []
+)
+
+# Testing odd (non-power-of-two for instance) shapes on purpose
+SHAPES = [
+    (384, 512),
+    (8, 384, 128),
+    (8, 784, 512),
+    (4, 16, 384),
+    (4, 16, 1024),
+    (2, 16, 2048),
+    (2, 16, 4096),
+    (1, 16, 12288),
+]
+
+BATCH = 4
+SEQ = 256
+EMBD = 16
+LATENT = 128
+DEVICES = [torch.device("cuda")]
+
+ACTIVATIONS = [
+    Activation.ReLU,
+    Activation.GeLU,
+    Activation.LeakyReLU,
+    Activation.SquaredReLU,
+    Activation.SmeLU,
+]
+
+
+@pytest.mark.skipif(
+    not xformers._is_functorch_available, reason="Functorch is not available"
+)
+@pytest.mark.skipif(not _gpu_available, reason="GPU is not available")
+@pytest.mark.parametrize("fused_pattern", FUSED_PATTERNS)
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("amp", [False, True])
+@pytest.mark.parametrize("bias", [False, True])
+@pytest.mark.parametrize("activation", ACTIVATIONS)
+@pytest.mark.parametrize("p", [0, 0.1, 0.5])
+@pytest.mark.parametrize("layer_norm_style", [LayerNormStyle.Pre, LayerNormStyle.Post])
+def test_nvfused_pattern_parity(
+    fused_pattern: nn.Module,
+    shape: tuple,
+    amp: bool,
+    bias: bool,
+    activation: Activation,
+    p: float,
+    layer_norm_style: LayerNormStyle,
+):
+
+    if (
+        fused_pattern != NVFusedBiasDropoutResLayerNorm
+        and layer_norm_style == LayerNormStyle.Post
+    ):
+        pytest.skip(
+            "Layer norm style doesn't apply, the same relevant params already tested once."
```

Comment:
How come we skip here? Should this error out if a user tries it?

Reply:
This is because the layernormstyle argument doesn't apply to the other patterns, so when layer_norm_style == LayerNormStyle.Pre all the possible configurations have already been tested once

Reply:
So should this just be `if fused_pattern != NVFusedBiasDropoutResLayerNorm` ?